reconcile/utils/expiration.py

```diff
@@ -1,41 +1,41 @@
 import datetime
+from typing import Iterable, Optional, Protocol, TypeVar, Union, cast
 
+DATE_FORMAT = "%Y-%m-%d"
+
+
+class FilterableRole(Protocol):
+    expiration_date: Optional[str]
 
-def has_valid_expiration_date(role: str) -> bool:
-    date_bool = True
-    if role is None:
-        return date_bool
-    else:
-        date_format = "%Y-%m-%d"
-        try:
-            date_bool = bool(datetime.datetime.strptime(role, date_format))
-        except ValueError:
-            date_bool = False
-        return date_bool
 
+DictsOrRoles = TypeVar(
+    "DictsOrRoles", bound=Union[Iterable[FilterableRole], Iterable[dict]]
+)
 
-def role_still_valid(role: str) -> bool:
-    if role is None:
-        return True
-    else:
-        exp_date = datetime.datetime.strptime(role, "%Y-%m-%d").date()
-        current_date = datetime.datetime.utcnow().date()
-        if current_date < exp_date:
-            return True
-        return False
 
+def date_expired(role: str) -> bool:
+    exp_date = datetime.datetime.strptime(role, DATE_FORMAT).date()
+    current_date = datetime.datetime.utcnow().date()
+    return current_date >= exp_date
 
-def filter(roles: list[dict], key: str = "expirationDate") -> list[dict]:
+
+def filter(roles: Optional[DictsOrRoles]) -> DictsOrRoles:
     """Filters roles and returns the ones which are not yet expired."""
     filtered = []
-    for r in roles:
-        expiration_date = r[key]
-        if not has_valid_expiration_date(expiration_date):
+    for r in roles or []:
+        if isinstance(r, dict):
+            key = "expirationDate"
+            expiration_date = r.get(key)
+        else:
+            key = "expiration_date"
+            expiration_date = r.expiration_date
+
+        try:
+            if not expiration_date or not date_expired(expiration_date):
+                filtered.append(r)
+        except ValueError:
             raise ValueError(
-                f"{key} field is not formatted as YYYY-MM-DD, "
-                f"currently set as {expiration_date}"
+                f"{key} field is not formatted as YYYY-MM-DD, currently set as {expiration_date}"
             )
-        if role_still_valid(expiration_date):
-            filtered.append(r)
 
-    return filtered
+    return cast(DictsOrRoles, filtered)
```

Comment:
Out of curiosity: would this work without a `cast`, by declaring

```
filtered: DictsOrRoles = []
```

above?

Reply:
That was my first thought as well but it didn't work because `DictsOrRoles` is `Iterable` and it doesn't support `append`. I tried it with `list` but then mypy throws this error :(

```
reconcile/quay_membership.py:97: error: <nothing> has no attribute "__iter__" (not iterable)  [attr-defined]
reconcile/quay_membership.py:97: error: Argument 1 to "filter" has incompatible type "Optional[List[RoleV1]]"; expected "None"  [arg-type]
```

Reply:
Interesting. By removing the `Iterable` from the type alias it seems to work locally for me. But I had to resort to `getattr()` again, because `Protocol` apparently doesnt like `isinstance` checks - its like turning in circles :sweat_smile: 

```python
DictOrRole = TypeVar(
    "DictOrRole", bound=Union[FilterableRole, dict]
)


def filter(roles: Optional[Iterable[DictOrRole]]) -> list[DictOrRole]:
    """Filters roles and returns the ones which are not yet expired."""
    filtered: list[DictOrRole] = []
    for r in roles or []:
        if isinstance(r, dict):
            key = "expirationDate"
            expiration_date = r.get(key)
        else:
            key = "expiration_date"
            expiration_date = getattr(r, key)

        try:
            if not expiration_date or not date_expired(expiration_date):
                filtered.append(r)
        except ValueError:
            raise ValueError(
                f"{key} field is not formatted as YYYY-MM-DD, currently set as {expiration_date}"
            )

    return filtered
```

Basically, Im trying to find a way to avoid `getattr()` and `cast()` because they are not ideal for catching stuff in static type checking phase. But I fail to find a way xD

Reply:
I think it would be a good idea to also test the dict path. Just in case, because this potentially affects quite a few integrations.

Reply:
I've edited my comment. Tested both, class-based with #2664 and the dict-based path with the current implementation of `quay-membership`

Reply:
Apparently this works:

```python
@runtime_checkable
class FilterableRole(Protocol):
    expiration_date: Optional[str]


DictOrRole = TypeVar(
    "DictOrRole", bound=Union[FilterableRole, dict]
)


def filter(roles: Optional[Iterable[DictOrRole]]) -> list[DictOrRole]:
    """Filters roles and returns the ones which are not yet expired."""
    filtered: list[DictOrRole] = []
    for r in roles or []:
        if isinstance(r, FilterableRole):
            key = "expiration_date"
            expiration_date = r.expiration_date
        else:
            key = "expirationDate"
            expiration_date = r.get(key)

        try:
            if not expiration_date or not date_expired(expiration_date):
                filtered.append(r)
        except ValueError:
            raise ValueError(
                f"{key} field is not formatted as YYYY-MM-DD, currently set as {expiration_date}"
            )

    return filtered
```

No `cast()`, no `getattr()`. Wdyt?

Reply:
Are you sure that `Iterable[DictOrRole]]` is semantic correct? `roles` should be either a list of `FilterableRole`s or a list of `dict`s, but `Iterable[DictOrRole]]` has mixed elements; am I right?

Reply:
I get this error:

```
reconcile/utils/expiration.py:32: error: Item "FilterableRole" of the upper bound "Union[FilterableRole, Dict[Any, Any]]" of type variable "DictOrRole" has no attribute "get"  [union-attr]
```

Reply:
> I get this error:

Note: I changed the branches of the if-statement. I think you still might have the old order, which uses `.get()` in the first `if` branch

Reply:
> Are you sure that Iterable[DictOrRole]] is semantic correct? roles should be either a list of FilterableRole's or a list of dict's, but Iterable[DictOrRole]] has mixed elements; am I right?

Good point. Yeah semantically this is flawed :sweat: